TaskSuspender utility

mihails-strasuns opened the pull request:
Based on idea by @daniel-zullo-sociomantic

Commit:
Don't resume suspended tasks on a shutdown

Commit:
Make debug trace of ThrottledTaskPool more useful

Usually one is only interested if resume/suspend was activated, not
intermediate numbers that led to it.

Commit:
Implement TaskSuspender utility

daniel-zullo commented on src/ocean/task/Scheduler.d:
```diff
@@ -660,6 +660,10 @@ final class Scheduler : IScheduler
+
+            // if calling last task resulted in a shutdown
+            if (this.state == State.Shutdown)
+                break;
```
would it be possible to add the `break` statement right after `task.kill()`? or could `this.resumeTask(task)` change the state of `this.state`?

mihails-strasuns replied:
Yep, it is for the latter - tesumed task calls shutdown internally and scheduler gets into shutdown state on the middle of processing suspend queue.

daniel-zullo replied:
right, then alternatively we could have

```D
        if (this.state != State.Shutdown)
            this.resumeTask(task);
                   
        // if calling last task resulted in a shutdown
        if (this.state == State.Shutdown)
        {
            task.kill();
            break;
        }
```
But I'm not sure that makes it better. It's up to you...

mihails-strasuns replied:
I have realized that original check is actually not needed at all - calling `shutdown` will kill all suspended tasks anyway. Updated to remove it completely.

daniel-zullo commented on src/ocean/task/util/TaskSuspender.d:
```diff
@@ -0,0 +1,87 @@
+
+    override public void resume ( )
+    {
+        theScheduler.delayedResume(task);
```
you might want to prepend `this.`  to task

daniel-zullo reviewed (commented):
I only have a cosmetic comment and a question and LGTM.
I'll test this patch in one application and report back. Thanks a lot!

daniel-zullo commented:
I've run a couple of tests and went all good.

Commit:
Remove redundant shutdown code from the scheduler

Suspended tasks are already killed by actual `shutdown` call and new
ones can't appear because `processEvents` kills itself when in shutdown
state.
